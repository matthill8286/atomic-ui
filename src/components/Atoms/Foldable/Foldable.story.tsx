import { withState } from '@dump247/storybook-state'
import { storiesOf } from '@storybook/react'
import * as React from 'react'
import { Foldable } from './Foldable'

const toggle = store => {
  store.set({ checkedValue: !store.state.checkedValue })
}

storiesOf('Design System/Atoms/Foldable', module).add(
  'Default',
  withState({ checkedValue: false }, store => (
    <React.Fragment>
      <Foldable isOpen={store.state.checkedValue}>HIDDEN</Foldable>
      <button onClick={() => toggle(store)}>Toggle Content</button>
    </React.Fragment>
  ))
)
