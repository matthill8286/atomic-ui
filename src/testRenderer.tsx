import { mount, ReactWrapper } from 'enzyme'
import React from 'react'
import { MemoryRouter as Router } from 'react-router-dom'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import { saiyanTheme } from '@/styles/sc-vars-saiyan'
import { Theme } from '@/types'

export function renderWithTheme<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>,
  theme = saiyanTheme
) {
  return renderer.create(<ThemeProvider theme={theme}>{tree}</ThemeProvider>).toJSON()
}

export function renderWithThemeAndRouter(tree: React.ReactElement, theme: Theme = saiyanTheme) {
  return renderer
    .create(
      <ThemeProvider theme={theme}>
        <Router>{tree}</Router>
      </ThemeProvider>
    )
    .toJSON()
}

export function renderWithRouter(tree: React.ReactElement) {
  return renderer.create(<Router>{tree}</Router>).toJSON()
}

export function mountWithTheme<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>
): ReactWrapper<P, S, C> {
  return mount(<ThemeProvider theme={saiyanTheme}>{tree}</ThemeProvider>)
}

export function mountWithThemeAndRouter<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>
): ReactWrapper<P, S, C> {
  return mount(
    <ThemeProvider theme={saiyanTheme}>
      <Router>{tree}</Router>
    </ThemeProvider>
  )
}
