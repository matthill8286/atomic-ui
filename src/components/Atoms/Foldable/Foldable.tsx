import React from 'react'
import { styled } from '@/styles/styled'
import { FoldableProps, FoldableState } from './Foldable.interface'

export const StyledFoldable = styled.div<{ looksOpenFromBreakpoint?: number }>`
  overflow: hidden;

  ${({ looksOpenFromBreakpoint }) =>
    looksOpenFromBreakpoint &&
    `
    @media (min-width: ${looksOpenFromBreakpoint}px) {
      height: auto !important;
    }
  `}
`

export class Foldable extends React.Component<FoldableProps, FoldableState> {
  private entryRef: React.RefObject<HTMLDivElement>
  private standardDuration: number
  private maxDuration: number

  constructor(props: FoldableProps) {
    super(props)
    this.entryRef = React.createRef()
    this.standardDuration = 300
    this.maxDuration = 1000

    this.state = {
      height: this._getHeight(),
      transition: `height ${this.standardDuration}ms linear`,
      looksOpenFromBreakpoint: props.looksOpenInitiallyFromBreakpoint,
    }
  }

  public componentDidMount() {
    if (this.props.isOpen) {
      this._updateHeightBeforeTransition()
    }
  }

  public componentDidUpdate(prevProps: FoldableProps) {
    if (this.props.isOpen !== prevProps.isOpen) {
      this._updateHeightBeforeTransition()
    }
  }

  public render() {
    const { isOpen: _isOpen, ...rest } = this.props
    const { looksOpenFromBreakpoint } = this.state
    return (
      <StyledFoldable
        data-test={'foldable'}
        style={this._applyTransition()}
        onTransitionEnd={this._afterTransition}
        className={this.props.className}
        looksOpenFromBreakpoint={looksOpenFromBreakpoint}
        {...rest}>
        <div data-test="inner-foldable" ref={this.entryRef}>
          {this.props.children}
        </div>
      </StyledFoldable>
    )
  }

  private _afterTransition = () => {
    const { isOpen } = this.props
    if (isOpen) {
      this.setState({
        height: 'auto',
      })
    }
  }

  private _scrollToElement(duration: number) {
    const currentElement = this.entryRef.current

    setTimeout(() => {
      const rect = currentElement?.getBoundingClientRect() || { top: 0 }
      if (rect.top < 0) {
        // reset viewport to calculate the right position
        window.scrollTo(0, 0)
        const rectUpdate = currentElement?.getBoundingClientRect() || { top: 0, height: 0 }
        window.scrollTo(0, rectUpdate.top - rectUpdate.height)
      }
    }, duration)
  }

  private _updateHeightBeforeTransition() {
    const { looksOpenFromBreakpoint } = this.state
    const isTransitionCancelled = !!looksOpenFromBreakpoint
    const currentElement = this.entryRef.current
    const calculatedDuration = this._getDuration()
    const animationType = this._getAnimationType(calculatedDuration)

    const height = isTransitionCancelled ? 'auto' : currentElement?.clientHeight ?? 'auto'
    const transition = isTransitionCancelled
      ? ''
      : `height ${calculatedDuration}ms ${animationType}`

    this.setState(
      {
        height,
        transition,
        looksOpenFromBreakpoint: undefined,
      },
      () => {
        if (!this.props.isOpen) {
          this._updateTransition()
        } else {
          this._scrollToElement(calculatedDuration)
        }
      }
    )
  }

  private _getDuration() {
    const { isOpen } = this.props
    const currentElement = this.entryRef.current
    const calcDuration = !!currentElement && currentElement.clientHeight / 2

    if (calcDuration) {
      if (!isOpen) {
        return this.standardDuration
      } else if (calcDuration > this.maxDuration) {
        return this.maxDuration
      } else if (calcDuration < this.standardDuration) {
        return this.standardDuration
      } else {
        return calcDuration
      }
    } else {
      return this.standardDuration
    }
  }

  private _getAnimationType(calculatedDuration: number) {
    const { isOpen } = this.props
    return calculatedDuration === this.standardDuration ? 'linear' : isOpen ? 'ease-in' : 'ease-out'
  }

  private _updateTransition() {
    const calculatedDuration = this._getDuration()
    const animationType = this._getAnimationType(calculatedDuration)
    this.setState({
      height: this._getHeight(),
      transition: `height ${calculatedDuration}ms ${animationType}`,
    })
  }

  private _applyTransition() {
    return {
      height: this.state.height,
      transition: this.state.transition,
    }
  }

  private _getHeight() {
    const { isOpen } = this.props
    return isOpen ? 'auto' : 0
  }
}
