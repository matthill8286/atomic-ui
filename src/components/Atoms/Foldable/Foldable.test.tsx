import { mount } from 'enzyme'
import React from 'react'
import { renderWithThemeAndRouter } from '@/testRenderer'
import { Foldable, StyledFoldable } from './Foldable'

const assertHeightIsCorrect = (wrapper, height) => {
  expect(wrapper.find(StyledFoldable).prop('style').height).toBe(height)
}

describe('Foldable', () => {
  beforeAll(() => {
    Object.defineProperty(HTMLElement.prototype, 'clientHeight', { configurable: true, value: 50 })
  })

  it('should render', () => {
    const tree = renderWithThemeAndRouter(
      <Foldable isOpen>
        <div>HIDDEN</div>
      </Foldable>
    )
    expect(tree).toMatchSnapshot()
  })
  it('should have a height of auto if initially opened', () => {
    const props = {
      isOpen: true,
    }
    const wrapper = mount(
      <Foldable {...props}>
        <div>HIDDEN</div>
      </Foldable>
    )
    expect(wrapper.state('height')).toBe(50)
    assertHeightIsCorrect(wrapper, 50)
  })

  it('should have a height of 0 if initially closed', () => {
    const props = {
      isOpen: false,
    }
    const wrapper = mount(
      <Foldable {...props}>
        <div>HIDDEN</div>
      </Foldable>
    )
    expect(wrapper.state('height')).toBe(0)
    assertHeightIsCorrect(wrapper, 0)
  })

  it('should update the height two times when opened', () => {
    const props = {
      isOpen: false,
    }
    const wrapper = mount<Foldable>(
      <Foldable {...props}>
        <div>HIDDEN</div>
      </Foldable>
    )
    const node = wrapper.find('[data-test="inner-foldable"]').instance()
    Object.defineProperty(node, 'clientHeight', { value: 100 })

    const spy = jest.spyOn(wrapper.instance(), 'setState')

    expect(wrapper.state('height')).toBe(0)
    assertHeightIsCorrect(wrapper, 0)
    expect(spy).toHaveBeenCalledTimes(0)

    wrapper.setProps({ isOpen: true })
    wrapper.childAt(0).simulate('transitionEnd')

    expect(spy).toHaveBeenCalledTimes(2)
    const calls = spy.mock.calls
    const call1Arg1 = calls && calls[0] && calls[0][0]
    const call2Arg1 = calls && calls[1] && calls[1][0]
    if (call1Arg1 && 'height' in call1Arg1) {
      expect(call1Arg1.height).toBe(100)
    }
    if (call2Arg1 && 'height' in call2Arg1) {
      expect(call2Arg1.height).toBe('auto')
    }
    spy.mockClear()
  })

  it('should update the height two times when closed', () => {
    const props = {
      isOpen: true,
    }
    const wrapper = mount<Foldable>(
      <Foldable {...props}>
        <div>HIDDEN</div>
      </Foldable>
    )
    const node = wrapper.find('[data-test="inner-foldable"]').instance()
    Object.defineProperty(node, 'clientHeight', { value: 100 })

    const spy = jest.spyOn(wrapper.instance(), 'setState')

    expect(wrapper.state('height')).toBe(50)
    assertHeightIsCorrect(wrapper, 50)
    expect(spy).toHaveBeenCalledTimes(0)

    wrapper.setProps({ isOpen: false })

    expect(spy).toHaveBeenCalledTimes(2)
    const calls = spy.mock.calls
    const call1Arg1 = calls && calls[0] && calls[0][0]
    const call2Arg1 = calls && calls[1] && calls[1][0]

    if (call1Arg1 && 'height' in call1Arg1) {
      expect(call1Arg1.height).toBe(100)
    }
    if (call2Arg1 && 'height' in call2Arg1) {
      expect(call2Arg1.height).toBe(0)
    }
    spy.mockClear()
  })
})
