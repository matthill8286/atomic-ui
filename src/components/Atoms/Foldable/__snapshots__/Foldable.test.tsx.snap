// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Foldable should render 1`] = `
.c0 {
  overflow: hidden;
}

<div
  className="c0"
  data-test="foldable"
  onTransitionEnd={[Function]}
  style={
    Object {
      "height": "auto",
      "transition": "height 300ms linear",
    }
  }
>
  <div
    data-test="inner-foldable"
  >
    <div>
      HIDDEN
    </div>
  </div>
</div>
`;
